agent/lib/kontena/websocket/client.rb

```diff
@@ -0,0 +1,762 @@
+    end
+
+    unless OpenSSL::SSL.verify_certificate_identity(ssl_cert, self.ssl_hostname)
+      raise Kontena::Websocket::SSLVerifyError.new(OpenSSL::X509::V_OK, ssl_cert, ssl_cert_chain), "Subject does not match hostname #{self.ssl_hostname}: #{ssl_cert.subject}"
+    end
+  end
+
+  # @param verify_result [Integer] OpenSSL::SSL::SSLSocket#verify_result
+  # @return [Kontena::Websocket::SSLVerifyError]
+  def ssl_verify_error(verify_result, ssl_cert = nil, ssl_cert_chain = nil)
+    ssl_verify_context = OpenSSL::X509::StoreContext.new(ssl_cert_store)
+    ssl_verify_context.error = verify_result
+
+    Kontena::Websocket::SSLVerifyError.new(ssl_verify_context.error, ssl_cert, ssl_cert_chain, ssl_verify_context.error_string)
+  end
+
+  # @return [OpenSSL::SSL::SSLContext]
+  def ssl_context
+    @ssl_context ||= OpenSSL::SSL::SSLContext.new().tap do |ssl_context|
+      ssl_context.set_params(**@ssl_params,
+        cert_store: self.ssl_cert_store,
+      )
+    end
+  end
+
+  # TODO: connect_deadline to impose a single deadline on the entire process
+  # XXX: specs
+  #
+  # @param ssl_socket [OpenSSL::SSL::SSLSocket]
+  # @raise [OpenSSL::SSL::SSLError]
+  # @raise [Kontena::Websocket::TimeoutError]
+  def ssl_connect(ssl_socket)
+    debug "ssl_connect..."
+    ret = ssl_socket.connect_nonblock
+  rescue IO::WaitReadable
+    debug "ssl_connect wait read: timeout=#{@connect_timeout}"
+    ssl_socket.wait_readable(@connect_timeout) or raise Kontena::Websocket::TimeoutError, "SSL connect read timeout after #{@connect_timeout}s"
+    retry
+  rescue IO::WaitWritable
+    debug "ssl_connect wait write: timeout=#{@connect_timeout}"
+    ssl_socket.wait_writable(@connect_timeout) or raise Kontena::Websocket::TimeoutError, "SSL connect write timeout after #{@connect_timeout}s"
+    retry
+  else
+    debug "ssl_connect: #{ret}"
+  end
+
+  # Connect to TCP server, perform SSL handshake, verify if required.
+  #
+  # @raise [Kontena::Websocket::ConnectError] from connect_tcp
+  # @raise [Kontena::Websocket::SSLConnectError]
+  # @raise [Kontena::Websocket::SSLVerifyError] errors that only happen with ssl_verify: true
+  # @return [OpenSSL::SSL::SSLSocket]
+  def connect_ssl
+    tcp_socket = self.connect_tcp
+    ssl_context = self.ssl_context
+
+    debug "connect_ssl: #{ssl_context.inspect} hostname=#{self.ssl_hostname}"
+
+    ssl_socket = OpenSSL::SSL::SSLSocket.new(tcp_socket, ssl_context)
+    ssl_socket.sync_close = true # close TCPSocket after SSL shutdown
+    ssl_socket.hostname = self.ssl_hostname # SNI
+
+    begin
+      self.ssl_connect(ssl_socket)
+    rescue OpenSSL::SSL::SSLError => exc
+      # SSL_connect returned=1 errno=0 state=error: certificate verify failed
+      if exc.message.end_with? 'certificate verify failed'
+        # ssl_socket.peer_cert is not set on errors :(
+        raise ssl_verify_error(ssl_socket.verify_result)
```

Comment:
The lack of `ssl_socket.peer_cert` on verify errors means that the resulting agent websocket client errors with `KONTENA_SSL_VERIFY=true` are less informative:
```
E, [2017-07-13T14:09:22.617530 #1] ERROR -- Kontena::WebsocketClient: unable to connect to SSL server with KONTENA_SSL_VERIFY=true: certificate verify failed: self signed certificate
```

Compared to the `KONTENA_SSL_VERIFY=` warning, when using `ssl_cert!` -> `ssl_verify_cert!`:

```
W, [2017-07-13T14:10:03.892937 #1]  WARN -- Kontena::WebsocketClient: insecure connection established with SSL errors: certificate verify failed: self signed certificate: /CN=kontena.test (issuer /CN=kontena.test)
```

It might be possible to include the cert information in these verify errors by replacing the built-in `verify_mode: OpenSSL::SSL::VERIFY_PEER` with the equivalent (?) verify checks that already happen in `ssl_verify_cert!`, but I'm not 100% sure yet that those are the same?